### PR TITLE
mgr: update daemon_state when necessary

### PR DESCRIPTION
**Pull request by cfanz**
Currently, mgr will refuse monitor's connection request if it doesn't have information about that monitor in its deamon_state. This will lead to a situation that the earlier started mgr cannot communicate well with the later added monitor.

Fixes: http://tracker.ceph.com/issues/37753
Signed-off-by: Xinyin Song <songxinying@sensetime.com>

### CODE REVIEW

File: src/mgr/Mgr.cc [hunk 1]
```diff
@@ -496,6 +502,35 @@ void Mgr::handle_service_map(MServiceMap *m)
   server.got_service_map();
 }
 
+void Mgr::handle_monmap(MMonMap *m)
+{
+  assert(lock.is_locked_by_me());
+  std::set<std::string> names_exist;
+
+  for (unsigned int i = 0; i < monmap.size(); i++) {
```
Comment (tserong) [inline src/mgr/Mgr.cc]: This won't build (`monmap` is an undeclared variable).

Comment (cfanz) [reply src/mgr/Mgr.cc]: Thanks for your review @tserong . The build error has been fixed now.

**Comment by cfanz** [comment]
With more tests, I find this pr still has a problem. 

In Monitor::refresh_from_paxos(), mon_metadata is still old value even thought monmap has increased. I've seen this problem in prior tests, and thought it was fixed by calling load_metadata() before refresh() method in each paxos_service. The tests at that time showed everything is ok, but now the weird thing occurs again.

I'm not familiar with monitor code, if anyone can give some advice, that will be a great appreciation.

**Comment by tserong** [comment]
@jecluis are you able to offer any advice here, or drag anyone else in? :-)

**Comment by cfanz** [comment]
I did some change on this pull request and it works fine on my test enviroment now. Here is how it works:
Mgr stores monitor metadata in Mgr::daemon_state variable, but this variable doesn't update when monmap changed. This will lead to some problem as described in [this issue](http://tracker.ceph.com/issues/37753).
The solution is let Mgr handle monmap message. One thing to notice is that when mgr receive a monmap message, it will request missed mon_metadata from monitor. But on this moment, Monitor::mon_metadata may have not been updated, so this request will fail. Since monitor will only send monmap to its subscribes only once for each epoch, mgr has to store a copy of the newest monmap, so that it can periodly check whether it can update its daemon_state. This periodly check is done in MgrStanby::tick().

File: src/mgr/Mgr.h [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
 #include "mds/FSMap.h"
 #include "messages/MFSMap.h"
+#include "messages/MMonMap.h"
```
Comment (tchaikov) [inline src/mgr/Mgr.h]: no need to `#include "messages/MMonMap.h"`, and we don't even need to forward declare `MMonMap`, what we need is a method `handle_mon_map()` which is called whenever a monmap is received. and monc will decode it for us.

File: src/mgr/Mgr.h [hunk 3]
```diff
@@ -85,6 +88,8 @@ class Mgr {
   void handle_osd_map();
   void handle_log(MLog *m);
   void handle_service_map(MServiceMap *m);
+  void handle_monmap(MMonMap *m);
```
Comment (tchaikov) [inline src/mgr/Mgr.h]: as explained above,
```c++
void handle_mon_map();
```
would suffice.

File: src/mgr/MgrStandby.cc [hunk 4]
```diff
@@ -161,6 +161,12 @@ int MgrStandby::init()
     client_messenger->wait();
     return r;
   }
+  /**
+   * this function must be called after mon.authenticate(); to avoid
+   * deadlock on MgrStandby::lock between MgrStandby::init() and
+   * MgrStandby::ms_dispatch()
+   */
```
Comment (tchaikov) [inline src/mgr/MgrStandby.cc]: could you elaborate this dead lock a little bit. is it related this change?

at the first glance, i don't really think this call is necessary for fixing a deadlock. it is added to prevent monclient from swallowing the `MMonMap`, so that other dispatcher(s) subscribing message events can be notified on it. in this case, it's `Mgr`.

**Comment by tchaikov** [comment]
also, you need to 

- summarize the change in the title of your commit message. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#title-of-pull-requests-and-title-of-commits 
- put the details in its body. 
- tag it appropriately, see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#tag-the-commit

please note, the first commit of this PR is different from the commit message. you need to revise it locally and repush for updating the latter.

File: src/mgr/MgrStandby.cc (hunk 4 above)
Comment (cfanz) [reply src/mgr/MgrStandby.cc]: At first, I thought monc.set_passthrough_monmap() is just setting a flag, so I added it next to monc.init(). The result is unexpected. I searched the code, but found nowhere is calling this function. With more debugs, I finally realize the reason and added this comment, hoping to help people avoid from the same trouble. But now it seems the comment brings more confuse than clarify. I prefer to remove this comment. What do you think?

Comment (tchaikov) [reply src/mgr/MgrStandby.cc]: @cfanz i don't think you answered my query. 

> could you elaborate this dead lock a little bit. is it related this change?

probably you could put down the time sequence of the dead lock? it's important to us. because, if this dead lock is not related to this change, we might need to consider a fix, and will need to backport it.

Comment (cfanz) [reply src/mgr/MgrStandby.cc]: Oh, sorry. Here is how the dead lock happens if we call monc.set_passthrough_monmap() before monc.authenticate():
After monc.authenticate() is called, there will be several rounds messages come and go between monitor and monc, monc will wait on its own lock before the authentication is finished. However, monitor will send a monmap to monc before monc finally finish authenticate. Since passthrough flag is set, mgr will also handle this monmap message in  MgrStandby::ms_dispatch. Because mgr is still holding its lock in MgrStandby::init() due to the un-returned monc.authenticate(), MgrStandy::ms_dispatch has to wait for the lock, this wait in return will block the whole messenger's process thread. However monc still needs the messenger to continue process its unfinished authenticate stuff. 
Then deadlock happens.

File: src/mgr/Mgr.h (hunk 2 above)
Comment (tchaikov) [reply src/mgr/Mgr.h]: i see. it's a little bit awkward. let me note down the problem in my own words:

assuming we have a cluster with mon.a and mon.b. and they have formed a quorum where mon.a is the leader. if we send "mon metadata" to monitor in `Mgr::handle_mon_map()`, we will have:

1. add mon.c to this cluster
2. mon.c sends a `MMonJoin` to mon.a for joining the cluster.
3. mon.a updates the monmap to include the new joiner
4. mgr.x is updated with the latest monmap
5. mgr.x tries to request for the metadata of mon.c, but since the monitors have not finished the election. the metadata are not updated yet. so mgr.x get a `-EINVAL`, see `Monitor::get_mon_metadata()
6. mon.c manages to join the quorum, and the metadata gets updated. see `Monitor::win_election()`. and monmap is not updated for the completion of election. and mgr.x won't ask the cluster for the updated metadata
6. mgr.x

File: src/mgr/Mgr.cc [hunk 5]
```diff
@@ -91,10 +92,16 @@ void MetadataUpdate::finish(int r)
         }
       }
 
+      // skip for monitor who doesn't have host-name yet
+      if (key.first == "mon" && daemon_meta.count("hostname") == 0) {
```
Comment (tchaikov) [inline src/mgr/Mgr.cc]: i'd suggest move this change into its own commit. and i think we should apply the sanity check to other daemons as well.

File: src/mgr/Mgr.h (hunk 2 above)
Comment (tchaikov) [reply src/mgr/Mgr.h]: again, no need to include `MMonMap.h`.

File: src/mgr/Mgr.cc [hunk 6]
```diff
@@ -496,6 +503,42 @@ void Mgr::handle_service_map(MServiceMap *m)
   server.got_service_map();
 }
 
+void Mgr::handle_mon_map(MMonMap *m)
+{
+  dout(20) << __func__ << dendl;
+  assert(lock.is_locked_by_me());
+  // read monmap from monclient instead of decoding from *m.
+  monc->with_monmap([this](const MonMap &m) { monmap = m; });
+  m->put();
```
Comment (tchaikov) [inline src/mgr/Mgr.cc]: no need to save the monmap here. i think we just need to cull the removed monitors from the `daemon_state`, like:

```c++
  std::set<std::string> names_exist;
  cluster_state.with_monmap([&] (auto &monmap) {
    for (unsigned int i = 0; i < monmap.size(); i++) {
      names_exist.insert(monmap.get_name(i));
    }
  });
  daemon_state.cull("mon", names_exist);
```

and let `DaemonServer::handle_report()` to take care of the new joiners.

File: src/mgr/MgrStandby.cc (hunk 4 above)
Comment (tchaikov) [reply src/mgr/MgrStandby.cc]: thanks! so it's only related to this change. that's a relief. i think you could rephrase it. like

> // only forward monmap updates after authentication finishes, otherwise
> // monc.authenticate() will be waiting for `MgrStandy::ms_dispatch()`
> // to acquire the lock forever, as it is already locked in the beginning of 
> // this method.

**Comment by cfanz** [comment]
> i'd suggest send `mon metadata` to monitor in `DaemonServer::handle_report()` if the report is rejected in hope to update the `daemon_state` registry if the sender is missing in it.

@tchaikov This idea is really great! Thus we don't even need to handle monmap. Since MgrClient will periodically call MgrClient::_send_stats(), so mgr server will finally update its daemon_state. But if we use this new plan, should we close this pr and open another or continually push on this? After all the new plan is so different from the current one.

File: src/mgr/Mgr.cc (hunk 5 above)
Comment (cfanz) [reply src/mgr/Mgr.cc]: > i'd suggest move this change into its own commit. and i think we should apply the sanity check to other daemons as well.

Yes, I agree that this check should be applied for every daemon type(mon,osd,mds,mgr). But  I don't quite understand about "move this change into its own commit". Could you explain more plain? Thanks.

**Comment by tchaikov** [comment]
@cfanz  i'd suggest reuse this PR. so we have more context for the succeeding discussion.

**Comment by tchaikov** [comment]
> Thus we don't even need to handle monmap

we could still cull the `daemon_state` with existing items in monmap.

File: src/mgr/Mgr.cc (hunk 5 above)
Comment (tchaikov) [reply src/mgr/Mgr.cc]: i mean we should split it into a separated commit.

**Comment by cfanz** [comment]
Now the strategy is:
1. When mgr receives an `MMgrReport` message from a monitor who doesn't in its `daemon_state`, mgr will try to fetch metadata about that monitor. Next time mgr receives `MMgrReport` from that monitor, it can establish a connection with that monitor.
2. Add `Mgr::handle_mon_map()` to cull `daemon_state` when monmap changes.

@tchaikov Could you please help review the new pushed codes? Thanks!

**Comment by tchaikov** [comment]
@tserong could you take another look?
